src/collapse/mod.rs

```diff
@@ -68,4 +83,304 @@ pub trait Collapse {
     /// - `Some(true)` means "yes, this implementation should work with this string"
     /// - `Some(false)` means "no, this implementation definitely won't work"
     fn is_applicable(&mut self, input: &str) -> Option<bool>;
+
+    #[cfg(test)]
+    fn set_nstacks_per_job(&mut self, n: usize);
+
+    #[cfg(test)]
+    fn set_nthreads(&mut self, n: usize);
+}
+
+/// Occurrences is a HashMap, which uses:
+/// * Fnv if single-threaded
+/// * CHashMap if multi-threaded
+#[derive(Clone, Debug)]
+enum Occurrences {
+    SingleThreaded(FnvHashMap<String, usize>),
+    MultiThreaded(Arc<CHashMap<String, usize>>),
+}
+
+impl Occurrences {
+    fn new_single_threaded() -> Self {
+        let map =
+            FnvHashMap::with_capacity_and_hasher(CAPACITY_HASHMAP, fnv::FnvBuildHasher::default());
+        Occurrences::SingleThreaded(map)
+    }
+
+    fn new_multi_threaded() -> Self {
+        let map = CHashMap::with_capacity(CAPACITY_HASHMAP);
+        let arc = Arc::new(map);
+        Occurrences::MultiThreaded(arc)
+    }
+
+    fn add(&mut self, key: String, count: usize) {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(map) => *map.entry(key).or_insert(0) += count,
+            MultiThreaded(arc) => arc.upsert(key, || count, |v| *v += count),
+        }
+    }
+
+    fn is_concurrent(&self) -> bool {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(_) => false,
+            MultiThreaded(_) => true,
+        }
+    }
+
+    fn write<W>(&mut self, mut writer: W) -> io::Result<()>
+    where
+        W: io::Write,
+    {
+        use self::Occurrences::*;
+        match self {
+            SingleThreaded(ref mut map) => {
+                let mut contents: Vec<_> = map.drain().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+            MultiThreaded(ref mut arc) => {
+                let map = match Arc::get_mut(arc) {
+                    Some(map) => map,
+                    None => panic!(
+                        "Attempting to drain the contents of a concurrent HashMap \
+                         when more than one thread has access to it, which is \
+                         not allowed."
+                    ),
+                };
+                let map = mem::replace(map, CHashMap::with_capacity(CAPACITY_HASHMAP));
+                let mut contents: Vec<_> = map.into_iter().collect();
+                contents.sort();
+                for (key, value) in contents {
+                    writeln!(writer, "{} {}", key, value)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+fn smallest_power_of_two_larger_than(n: usize) -> usize {
```

Comment:
I think this can just use `usize::next_power_of_two`. If not, it should be renamed.